pkg/daemon/config_drift_monitor.go

```diff
@@ -0,0 +1,378 @@
+}
+
+// Holds a single Config Drift Watcher and starts / stops it as necessary while
+// ensuring that only a single Config Drift Monitor is running at any given
+// time (within its scope).
+func NewConfigDriftMonitor() ConfigDriftMonitor {
+	return &configDriftMonitor{
+		mu:     sync.Mutex{},
+		stopCh: make(chan struct{}),
+	}
+}
+
+// Determines if a Config Drift Watcher is running; useful for avoiding certain
+// logic such as emitting a startup event.
+func (c *configDriftMonitor) IsRunning() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.cdw != nil
+}
+
+// Starts the Config Drift Monitor if not already started. It will create a new
+// Config Drift Monitor instance, if necessary.
+func (c *configDriftMonitor) Start(opts ConfigDriftMonitorOpts) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cdw != nil {
+		return nil
+	}
+
+	var err error
+
+	// We don't have a Config Drift Watcher running, create one.
+	c.cdw, err = newConfigDriftWatcher(opts)
+	if err != nil {
+		return err
+	}
+
+	c.cdw.start()
+
+	return nil
+}
+
+// Provides a channel for listening if the currently running Config Drift
+// Monitor has been stopped. Can be used similar to <-ctx.Done().
+func (c *configDriftMonitor) Done() <-chan struct{} {
+	out := make(chan struct{})
+
+	go func() {
+		select {
+		case <-c.stopCh:
+			out <- struct{}{}
+			close(out)
+			return
+		}
+	}()
+
+	return out
+}
+
+// Stops the currently running Config Drift Monitor and deletes its instance so
+// that it cannot be reused. Also sends a signal to the Done() listener.
+func (c *configDriftMonitor) Stop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cdw != nil {
+		c.cdw.stop()
+		c.cdw = nil
+		c.stopCh <- struct{}{}
+	}
+}
+
+// Creates a new config drift watcher for a given machineconfig and registers
+// a callback for when a drift occurs.
+func newConfigDriftWatcher(opts ConfigDriftMonitorOpts) (*configDriftWatcher, error) {
+	if opts.OnDrift == nil {
+		return nil, fmt.Errorf("no ondrift function attached")
+	}
+
+	if opts.ErrChan == nil {
+		return nil, fmt.Errorf("no error channel provided")
+	}
+
+	if opts.MachineConfig == nil {
+		return nil, fmt.Errorf("no machine config provided")
+	}
+
+	if opts.SystemdPath == "" {
+		opts.SystemdPath = pathSystemd
+	}
+
+	c := &configDriftWatcher{
+		ConfigDriftMonitorOpts: opts,
+		stopCh:                 make(chan struct{}),
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, fmt.Errorf("could not initialize config drift monitor: %w", err)
+	}
+
+	return c, nil
+}
+
+// Performs the initial setup and initialization of the Config Drift Monitor,
+// such as identifying files from the MachineConfig, wiring up the watchers,
+// etc.
+func (c *configDriftWatcher) initialize() error {
+	var err error
+
+	// Initialize the fsnotify watcher
+	c.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("could not start watcher: %w", err)
+	}
+
+	glog.V(4).Infof("Initializing Config Drift Monitor")
```

Comment:
are there any glog.Info level config drift monitor running messages? i see starting, i see initializing, i see stopping but (and my search might be bad here) is there anything that just explicitly says yep it's running now:"

Reply:
No, but it would be a good idea to have them!